### DISCUSSION TITLE
Questions on how to contribute

### DISCUSSION THREAD

**Discussion by Thelvaen** [bdunderscore/modular-avatar#997] [General]
Hello,

I'm trying to convert Face Tracking prefabs to MA, as for a lot of reasons I don't want to use VRCFury, I could do the installation manually, but it gets old really fast, and let's be honest here, drag'n'drop the prefab on the avatar is still a lot more comfortable.

To that end, and for other projects I'm in the process of creating a couple of MA components that are needed:
* Blendshape syncer: gets a string as input for the source/remote SkinnedMeshRenderer, and sync all the weight with the local SkinnedMeshRenderer (already have that one working, just haven't done an editor the way you're doing it yet),

* Parameters Installer: gets a parameter asset file as input, and installs all the parameters as they are specified (I kinda think you're already working on that one according to issues I've read, but automated translation being what they are, I'm not entirely sure),

Would you welcome those features?

**Comment by bdunderscore** [comment]
- Regarding Blendshape Syncer - what do you mean by getting a string as input? Is the change just to add a mode to sync all blendshapes instead of just specified ones?
- Re Parameters Installer - this is planned, in one form or another. The MA parameters component has some additional configuration beyond the base VRCSDK parameters, so I'm considering just having a way to populate it from a parameters asset.

**Comment by Thelvaen** [comment]
Yes, for the blendshape, it's adding a mode (for now I made it with an extra component) that sync all blendshape with the same name on both SkinnedMeshRenderer. I know it will not work for viseme related stuff, but it is needed for example for the Face Tracking Debug object.

**Comment by bdunderscore** [comment]
Since there’s already a blendshape sync component, it’s probably best to extend that with a new sync-all mode rather than adding a new similar component…

**Comment by Thelvaen** [comment]
Yup, definitely, since the way I'm doing it right now is through replicating your component but with a different rebind logic, it should be pretty easy to just integrate that in the existing one, and change the editor to only display the settings depending on the mode.

**Comment by Thelvaen** [comment]
So I have the editor & modified component working.
I was thinking of going for a bool toggle named "Simple Mode" (or something like that), where in Simple Mode it will just sync all blendshapes, and if you untick the Simple Mode box, it will display the current blendshape list.
Or do you prefer going for an Enum style drop down for the mode selection?

**Comment by bdunderscore** [comment]
Controls should generally describe what they do - so it should be a "Sync all blendshapes" checkbox, not a mysterious "simple mode".

**Comment by Thelvaen** [comment]
Would that match what you have in mind for that feature?

![image](https://github.com/user-attachments/assets/f9a19d79-1111-4be8-a611-87d65014806d)
![image](https://github.com/user-attachments/assets/35b56375-1fb7-472f-955c-10c10a7d0215)

If we're on the same page, I'll do a PR once I made a couple of extra tests at home (want to test a couple of edge cases and probably provide a way to detect the in prefab editing for asset maker).

**Comment by bdunderscore** [comment]
For source renderer please use an `AvatarObjectReference` through a property field as this will handle making a path reference look like an object selection interface automatically (take a look at how the list is currently working for an example)

**Comment by Thelvaen** [comment]
Idea is also to provide a solution that would allow people to distribute pre-configured prefab, without avatar (like the Face Tracking one for example).

**Comment by bdunderscore** [comment]
Yes, that's the idea behind AvatarObjectReference - it binds to a path, but it also does a bunch of UI tricks to make it work more smoothly.

**Comment by Thelvaen** [comment]
So I did try to replace the String with an AvatarObjectReference, it does not allow to get the path written manually.

Under the hood, my code does use this type once the proper SkinnedMeshRenderer has been found

```csharp
                var AvatarRoot = RuntimeUtil.FindAvatarTransformInParents(this.transform);
                if (AvatarRoot == null) return;
                var renderers = AvatarRoot.gameObject.GetComponentsInChildren<SkinnedMeshRenderer>();
                if (renderers.Length == 0)
                {
                    throw new ArgumentException("No skinned mesh renderer found");
                }
                this.sourceRenderer = null;
                foreach (SkinnedMeshRenderer renderer in renderers)
                {
                    GameObject gameObject = renderer.gameObject;
                    if (gameObject.name.Equals(this.SourceRendererName))
                    {
                        this.sourceRenderer = renderer;
                        break;
                    }
                    if (this.sourceRenderer == null)
                    {
                        throw new ArgumentException("Can't find a Skinned Mesh with name " + this.SourceRendererName);
                    }
                }

                var sourceCount = this.sourceRenderer.sharedMesh.blendShapeCount;
                var targetCount = localRenderer.sharedMesh.blendShapeCount;

                // creating the AvatarObjectReference object
                var sourceReference = new AvatarObjectReference();
                sourceReference.Set(this.sourceRenderer.gameObject);
                GameObject Test = sourceReference.Get(this.sourceRenderer);
                if (Test != this.sourceRenderer.gameObject) return;
                
                for (int i = 0; i < sourceCount; i++)
                {
                    for (int j = 0; j < targetCount; j++)
                    {
                        if (this.sourceRenderer.sharedMesh.GetBlendShapeName(i).Equals(localRenderer.sharedMesh.GetBlendShapeName(j)))
                        {
                            _editorBindings.Add(new EditorBlendshapeBinding()
                            {
                                TargetMesh = this.sourceRenderer,
                                RemoteBlendshapeIndex = i,
                                LocalBlendshapeIndex = j
                            });
                            Bindings.Add(new BlendshapeBinding()
                            {
                                ReferenceMesh = sourceReference,
                                Blendshape = this.sourceRenderer.sharedMesh.GetBlendShapeName(i),
                                LocalBlendshape = this.sourceRenderer.sharedMesh.GetBlendShapeName(i)
                            });
                        }
                    }
                }                    
```

**Comment by bdunderscore** [comment]
Generally you’d put an AvatarObjectReference in your component as a serialized field, there’s not much point creating it at runtime…

**Comment by bdunderscore** [comment]
https://github.com/bdunderscore/modular-avatar/blob/e2e0b88dfab16d80f3d3932fc838bed0c59c76fc/Runtime/ModularAvatarMeshSettings.cs#L23 example of use in a component

**Comment by Thelvaen** [comment]
The problem with using AvatarObjectReference is that it does not work properly when you're working on your prefab in the prefab editor, it needs to be in an avatar context with the valid path to the object.

You can see on the screenshots that all references are disappearing because of that invalid path, despite being properly in the object when I drag'n'drop the prefab on an avatar.

![image](https://github.com/user-attachments/assets/1cf3754c-ec22-48fc-a4ea-6c37f236031e)
![image](https://github.com/user-attachments/assets/d3ba5cb6-4d4b-49cb-aec8-eceecef66ffb)
![image](https://github.com/user-attachments/assets/406d5897-1819-4946-9f40-0b6655a3abbd)

**Comment by bdunderscore** [comment]
Your screenshots look like the references are there…?

I think it’s important for MA components to be consistent with each other, which is why I’m resistant to the idea of a bespoke way of referencing an object here. What’s the specific workflow you’re envisioning here?

**Comment by bdunderscore** [comment]
Converting to discussion as there’s too many things mixed up in one thread for an issue

**Comment by Thelvaen** [comment]
I'm trying to convert the Face Tracking VRCF prefabs in MA prefabs. And I'm trying to replicate the way it can be installed on Avatar, with minimal End User interactions.

The way it's currently done through VRCF, is, you take the prefab, drop it on your Avatar, hit the build button, and it works, you don't need to configure anything on any of the components of the prefab.

To achieve that, it's using String paths to reference target objects, and it will find said object on Runtime, that way it can minimize the end user interaction with the prefab (when I remember some of the support request I got in my previous jobs, I can totally understand why you'd want to minimize end user operations to the strict minimum if possible).

So being able to check that the prefab you're building is correct, and has all the correct references, outside of a scene, before packaging it/pushing it to Github feels like something that needs to be done.

**Comment by bdunderscore** [comment]
So, that sounds like what AvatarObjectReference does - it stores a path from the avatar root, and uses that to locate a target object. I'm just a bit confused about what part of your workflow isn't working - it sounds like maybe there's some UI/UX issues with the AOR editor, in which case we should just fix that.

**Comment by bdunderscore** [comment]
![image](https://github.com/user-attachments/assets/9b288085-46ce-4cce-aae5-71336d503989)

Here's a blendshape sync (using AOR for the mesh) in isolation mode. We could change the property drawer to allow the mesh path to be edited when outside of an avatar. (I also note that the + button breaks when outside of an avatar, which probably should be fixed)
